exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,203 @@
+    ///
+    /// *This method should not use external data, that is, it must be a pure function.*
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[macro_use] extern crate exonum;
+    /// #
+    /// use exonum::blockchain::Transaction;
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::messages::Message;
+    /// # use exonum::blockchain::TransactionStatus;
+    /// # use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #       const SIZE = 32;
+    ///         field public_key:    &PublicKey    [0 => 32]
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn verify(&self) -> bool {
+    ///         self.verify_signature(self.public_key())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn execute(&self, _: &mut Fork) -> TransactionStatus { TransactionStatus::Succeeded }
+    /// }
+    /// # fn main() {}
+    fn verify(&self) -> bool;
+
+    /// Takes the current blockchain state via `fork` and can modify it if certain conditions
+    /// are met.
+    ///
+    /// # Notes
+    ///
+    /// - When programming `execute`, you should perform state-related checks before any changes
+    /// to the state and return early if these checks fail.
+    /// - If the execute method of a transaction raises a `panic`, the changes made by the
+    /// transactions are discarded, but the transaction itself is still considered committed.
+    /// - A transaction execution status (see `ExecutionStatus` for the details) is stored in the
+    /// blockchain and can be accessed through api.
+    ///
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[macro_use] extern crate exonum;
+    /// #
+    /// use exonum::blockchain::{Transaction, TransactionStatus};
+    /// use exonum::crypto::PublicKey;
+    /// use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #       const SIZE = 32;
+    /// #       field public_key:    &PublicKey    [0 => 32]
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     fn execute(&self, fork: &mut Fork) -> TransactionStatus {
+    ///         // Read and/or write into storage.
+    ///         // ...
+    ///
+    ///         // Returns execution status.
+    ///         TransactionStatus::Succeeded
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn verify(&self) -> bool { true }
+    /// }
+    /// # fn main() {}
+    fn execute(&self, fork: &mut Fork) -> TransactionStatus;
+}
+
+/// Execution status of the transaction.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionStatus {
```

Comment:
An alternative would be to use 

```Rust
type TransactionStatus = Result<(), TransactionError>;

enum TransactionError {
    Panic,
    UnknownFailure,
    Code(u8),
}
```

Reply:
Do we want to use `?` when working with transaction? Do we want clients to use `?` when defining transactions? Do we want to allow clients to define `Into<TransactionError> for MyCustomClientError` ?

Reply:
`Result` is nice, but I don't like `()` as `Ok` value: someone can want to have different statuses for the successful execution. What do you think about `Result<u8, TransactionError>`? Not sure about "raw" `u8` though.

Reply:
Ah, so `Custom` in `TransactionStatus` can work for both errors and successes?

Reply:
Yes, I had that idea. It isn't ideal, but it should be flexible enough for different potential use-cases.

Reply:
That means, that we can't really say if the transaction was successful or not, at the core level?

So, the fully-general(but still code-based) type would look like this?

```Rust
type TransactionResult = Result<TransactionValue, TransactionError>;

struct TransactionValue(u8);
const Success: TransactionVaue = TransactionValue(0);

enum TransactionError {
    Panic,
    UnknownFailure,
    Code(u8)
}
```

I don't really know the full set of tradeoffs here, so I don't know which solution is better. Allowing to use `?` when defining transactions seems somewhat nice though?

Reply:
I agree that `?` is really convenient, but I am not sure about such `TransactionValue` implementation. Maybe it makes sense to use `enum` too? Something like that:
```rust
type TransactionResult = Result<TransactionValue, TransactionError>;

enum TransactionValue {
    Success,
    Code(u8),
}

enum TransactionError {
    Panic,
    UnknownFailure,
    Code(u8)
}
```

Reply:
@DarkEld3r yep, using `enum` for TransactionValue looks great! We might want to add `#[doc(hidden)] __non_exhaustive` if we think that we might want to extend `u8` to something else in the future.